### PR TITLE
Mob-3640 add loyalty program APIs

### PR DESCRIPTION
**Pull request by mionescu95**
## JIRA
[MOB-3640](https://karhoo.atlassian.net/browse/MOB-3640)

## DESCRIPTION
Added Loyalty APIs

## Types of changes

What types of changes does your code introduce to the project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix 
- [x] New feature 
- [ ] Tech Debt
- [ ] Breaking change 
- [ ] Documentation Update
- [ ] Other: *define what type of change it is*

## CHECKLIST

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging code._
Please review the guidelines for contributing to this repository.

- [ ] All work relating to the ticket is complete (Code complete/Acceptance Criteria met)
- [ ] Lint and unit tests pass locally (if appropriate)
- [ ] Added tests that prove the fix is effective or that feature works (if appropriate)
- [ ] Documentation (if appropriate)

## SCREENSHOTS
[Add screenshots if any for changes you have made to the App/UI]

### CODE REVIEW

File: sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyBurnPointsInteractor.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
+
+internal class LoyaltyBurnPointsInteractor @Inject constructor(credentialsManager: CredentialsManager,
+                                                           private val apiTemplate: APITemplate,
+                                                           context: CoroutineContext = Dispatchers.Main
+                                                          ) :
+    BaseCallInteractor<LoyaltyPoints>(true, credentialsManager, apiTemplate, context) {
```
Comment (mateid-kr) [inline sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyBurnPointsInteractor.kt:19]: Boolean literal argument is without parameter name. Maybe add the param name also:
requestRequiresToken = true

File: sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyBurnPointsInteractor.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    internal var loyaltyId: String? = null
```
Comment (mateid-kr) [inline sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyBurnPointsInteractor.kt:21]: the properties can be made private, don't think there's a reason to be made internal.

File: sdk/src/main/java/com/karhoo/sdk/api/network/client/APITemplate.kt [hunk 3]
```diff
@@ -235,6 +246,25 @@ interface   APITemplate {
...
+    @GET(LOYALTY_STATUS)
+    fun loyaltyStatus(@Path(IDENTIFIER_ID) id : String): Deferred<Resource<LoyaltyStatus>>
+
+    @GET(LOYALTY_BURNPOINTS)
+    fun loyaltyBurnPoints(@Path(IDENTIFIER_ID) id: String, @Path(IDENTIFIER_CURRENCY) currency:
+    String, @Query(IDENTIFIER_AMOUNT) amount: Int): Deferred<Resource<LoyaltyPoints>>
+
+    @GET(LOYALTY_EARNPOINTS)
+    fun loyaltyPointsToEarn(@Path(IDENTIFIER_ID) id: String, @Path(IDENTIFIER_CURRENCY) currency:
+    String, @Query(IDENTIFIER_TOTALAMOUNT) totalAmount: Int, @Query(IDENTIFIER_BURNPOINTS)
```
Comment (mateid-kr) [inline sdk/src/main/java/com/karhoo/sdk/api/network/client/APITemplate.kt:258]: The identation seems broken here, we should have params on separate lines for easier reading. Same applies to above

File: sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyBurnPointsInteractor.kt (hunk 2 above)
Comment (mateid-kr) [reply sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyBurnPointsInteractor.kt:21]: They can be passed in from the constructor

File: sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyPointsToEarnInteractor.kt [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import com.karhoo.sdk.api.network.client.APITemplate
+import com.karhoo.sdk.api.network.response.Resource
+import com.karhoo.sdk.api.service.common.BaseCallInteractor
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
+
+internal class LoyaltyPointsToEarnInteractor @Inject constructor(credentialsManager: CredentialsManager,
```
Comment (mateid-kr) [inline sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyPointsToEarnInteractor.kt:15]: I have mixed feelings about having two separate interactors, we could have one which handles both earning and burning scenarios 
@mohadian what do you think?

File: sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyBurnPointsInteractor.kt (hunk 2 above)
Comment (mateid-kr) [reply sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyBurnPointsInteractor.kt:21]: Same applies for the below properties in the LoyaltyPointsToEarnInteractor

**Review by mateid-kr — CHANGES_REQUESTED** [review]

File: sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyStatusInteractor.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import com.karhoo.sdk.api.network.client.APITemplate
+import com.karhoo.sdk.api.network.response.Resource
+import com.karhoo.sdk.api.service.common.BaseCallInteractor
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import javax.inject.Inject
+import kotlin.coroutines.CoroutineContext
+
+internal class LoyaltyStatusInteractor @Inject constructor(credentialsManager: CredentialsManager,
```
Comment (mateid-kr) [inline sdk/src/main/java/com/karhoo/sdk/api/service/loyalty/LoyaltyStatusInteractor.kt:15]: There are no unit tests for these interactors / requests

File: sdk/src/main/java/com/karhoo/sdk/api/network/request/LoyaltyPreAuth.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package com.karhoo.sdk.api.network.request
+
+import android.os.Parcelable
+import com.google.gson.annotations.SerializedName
+import kotlinx.android.parcel.Parcelize
+
+@Parcelize
+data class LoyaltyPreAuth(@SerializedName("currency") val currency: String? = null,
```
Comment (mohadian) [inline sdk/src/main/java/com/karhoo/sdk/api/network/request/LoyaltyPreAuth.kt:8]: the ios LoyaltyPreAuth contains 
`public let identifier: String`

I think the identifier can be removed in the iOS version (Android is fine?)

File: sdk/src/main/java/com/karhoo/sdk/api/network/request/LoyaltyPreAuth.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+                       @SerializedName("points") val points: Int? = null,
+                       @SerializedName("flexpay") val flexpay: Boolean? = null,
+                       @SerializedName("loyalty_membership") val loyaltyMembership: String? = null) : Parcelable
```
Comment (mohadian) [inline sdk/src/main/java/com/karhoo/sdk/api/network/request/LoyaltyPreAuth.kt:11]: in the iOS, it is called `membership: String`

**Review by mateid-kr — APPROVED** [review]

**Pull request merged**